src/SourceBuild/content/eng/tools/BinaryToolKit/BinaryTool.CompareBinariesAgainstBaselines.cs

```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.FileSystemGlobbing;
+using Microsoft.Extensions.Logging;
+
+namespace BinaryToolKit;
+
+public partial class BinaryTool
+{
+    private const string PatternsToRemoveMarker = "# Consider removing the following unused patterns from the baseline file:";
+
+    private List<string> CompareBinariesAgainstBaselines(IEnumerable<string> detectedBinaries)
+    {
+        Log.LogInformation("Comparing detected binaries to baseline(s).");
+
+        var binariesToRemove = GetUnmatchedBinaries(detectedBinaries, AllowedBinariesFile, UpdatedAllowedBinariesFile).ToList();
+
+        if (ModeOption != Mode.ModeOptions.clean)
+        {
+            var nonSbBinariesToRemove = GetUnmatchedBinaries(detectedBinaries, DisallowedSbBinariesFile, UpdatedDisallowedSbBinariesFile).ToList();
+            var newBinaries = binariesToRemove.Intersect(nonSbBinariesToRemove);
+
+            if (newBinaries.Any())
+            {
+                Log.LogWarning($"    {newBinaries.Count()} new binaries detected. Check {NewBinariesFile}");
+                File.WriteAllLines(NewBinariesFile!, newBinaries);
+            }
+        }
+
+        Log.LogInformation("Finished comparing binaries.");
+
+        return binariesToRemove;
+    }
+
+    private IEnumerable<string> GetUnmatchedBinaries(IEnumerable<string> searchFiles, string? baselineFile, string? updatedBaselineFile)
+    {
+        var patterns = ParseBaselineFile(baselineFile);
+
+        if (ModeOption == Mode.ModeOptions.clean)
+        {
+            Matcher matcher = new Matcher(StringComparison.Ordinal);
+            matcher.AddInclude("**/*");
+            matcher.AddExcludePatterns(patterns);
+
+            return matcher.Match(TargetDirectory, searchFiles).Files.Select(file => file.Path);
+        }
+        else
+        {
+            HashSet<string> unusedPatterns = new HashSet<string>(patterns);
+            HashSet<string> unmatchedFiles = new HashSet<string>(searchFiles);
+
+            foreach (string pattern in patterns)
+            {
+                Matcher matcher = new Matcher(StringComparison.Ordinal);
+                matcher.AddInclude(pattern);
+                
+                var matches = matcher.Match(TargetDirectory, searchFiles);
+                if (matches.HasMatches)
+                {
+                    unusedPatterns.Remove(pattern);
+                    unmatchedFiles.ExceptWith(matches.Files.Select(file => file.Path));
+                }
+            }
+
+            UpdateBaselineFile(baselineFile!, updatedBaselineFile!, unusedPatterns);
```

Comment:
What happens if `baselineFile` or `updatedBaselineFile` are null here? If you know for sure they are not null then the type should reflect that.

Reply:
My bad, they can be null here. the `!`s were leftover by accident.